bpf/protocol_common.h

```diff
@@ -95,114 +97,189 @@ static __always_inline http_connection_metadata_t *connection_meta_by_direction(
-				void *ubuf;
-			};
-			size_t count;
-		};
-	};
+struct iov_iter___v64 {
+    u8 iter_type;
+    bool nofault;
+    bool data_source;
+    size_t iov_offset;
+    union {
+        struct iovec __ubuf_iovec;
+        struct {
+            union {
+                /* use iter_iov() to get the current vec */
+                const struct iovec *__iov;
+                const struct kvec *kvec;
+                const struct bio_vec *bvec;
+                struct xarray *xarray;
+                void *ubuf;
+            };
+            size_t count;
+        };
+    };
     union {
-		unsigned long nr_segs;
-		loff_t xarray_start;
-	};
+        unsigned long nr_segs;
+        loff_t xarray_start;
+    };
 };
 
+struct iov_iter___v60 {
+    u8 iter_type;
+    bool nofault;
+    bool data_source;
+    bool user_backed;
+    union {
+        size_t iov_offset;
+        int last_offset;
+    };
+    size_t count;
+    union {
+        const struct iovec *iov;
+        const struct kvec *kvec;
+        const struct bio_vec *bvec;
+        struct xarray *xarray;
+        struct pipe_inode_info *pipe;
+        void *ubuf;
+    };
+    union {
+        unsigned long nr_segs;
+        struct {
+            unsigned int head;
+            unsigned int start_head;
+        };
+        loff_t xarray_start;
+    };
+};
 
-static __always_inline int read_msghdr_buf(struct msghdr *msg, u8* buf, int max_len) {
-    struct iov_iter msg_iter = BPF_CORE_READ(msg, msg_iter);
-    u8 msg_iter_type = 0;
-
-    if (bpf_core_field_exists(msg_iter.iter_type)) {
-        bpf_probe_read(&msg_iter_type, sizeof(u8), &(msg_iter.iter_type));
-        bpf_dbg_printk("msg iter type exists, read value %d", msg_iter_type);
-    }
-
-    bpf_dbg_printk("iter type %d", msg_iter_type);
 
-    struct iovec *iov = NULL;
+// older struct that features 'type' instead of 'iter_type'
+struct iov_iter___v58 {
+    unsigned int type;
+    size_t iov_offset;
+    size_t count;
+    union {
+        const struct iovec *iov;
+        const struct kvec *kvec;
+        const struct bio_vec *bvec;
+        struct pipe_inode_info *pipe;
+    };
+    union {
+        unsigned long nr_segs;
+        struct {
+            unsigned int head;
+            unsigned int start_head;
+        };
+    };
+};
 
-    u32 l = max_len;
-    bpf_clamp_umax(l, IO_VEC_MAX_LEN);
+// helper struct used by get_iovec_ctx
+struct iovec_iter_ctx {
+    unsigned int iter_type;
+    size_t iov_offset;
+    size_t count;
+    unsigned long nr_segs;
+    const struct iovec *iov;
+    const void *ubuf;
+};
 
-    if (bpf_core_field_exists(msg_iter.iov)) {
-        bpf_probe_read(&iov, sizeof(struct iovec *), &(msg_iter.iov));
-        bpf_dbg_printk("iov exists, read value %llx", iov);
+// extracts kernel specific iov_iter information into a iovec_iter_ctx instance
+static __always_inline void get_iovec_ctx(struct iovec_iter_ctx* ctx, struct msghdr *msg) {
+    if (LINUX_KERNEL_VERSION <= KERNEL_VERSION(5, 13, 0)) {
```

Comment:
Relying on runtime checks as done by the previous commits of this PR started to impact code clarity. Instead of checking whether enums and fields exist, I decided to streamline everything solely based on the kernel version. Whilst the downside is a bit of repetition/boilerplate, this makes it very obvious what's being used in which kernel version, and is all encapsulated by this function. The code is now linear, rather than a convoluted decision tree.

I've chosen to keep the flavored struct naming scheme (`___vXX`) as it's very explicit about the expected kernel version, and it still plays nice in case we decide to CORE, as the earlier commits did.